### ISSUE TITLE
Support multiple possible schema for validation

### ISSUE THREAD

**Issue by apupier** [microsoft/vscode-json-languageservice#60]
Currently, if several schemas are possible a validation warning is provided:
`Matches multiple schemas when only one must validate.`

https://github.com/microsoft/vscode-json-languageservice/blob/fad098244642100e5054dedefed9da905b24aedc/src/parser/jsonParser.ts#L454-L459

I can be wrong but it seems a possible case. it is causing a false-positive in UI.
I think that it would be nice to validate against the various potential schemas and if one is matching, to not throw the validation.

Does it sounds reasonable?

Or it is the schema should never hit this case. In this case, how to avoid that in a clean way on schema side?

**Comment by SPGoding** [comment]
Use `anyOf` instead of `oneOf` in your schema can solve this if I understand correctly.

**Comment by apupier** [comment]
> Use `anyOf` instead of `oneOf` in your schema can solve this if I understand correctly.

good point.
based on http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.2.1.2

<pre><code>
9.2.1.2. anyOf
This keyword's value MUST be a non-empty array. Each item of the array MUST be a valid JSON Schema.

An instance validates successfully against this keyword if it validates successfully against <b>at least one schema</b> defined by this keyword's value. Note that when annotations are being collected, all subschemas MUST be examined so that annotations are collected from each subschema that validates successfully.

9.2.1.3. oneOf
This keyword's value MUST be a non-empty array. Each item of the array MUST be a valid JSON Schema.

An instance validates successfully against this keyword if it validates successfully against <b>exactly one schema</b> defined by this keyword's value.
</pre>

so the fact that a validation error is provided respect the specification. What would be interesting would be to improve the error message to point to the specification maybe?

nota: this will surely break a lot of other softwares that rely on this message: https://github.com/search?q=%22Matches+multiple+schemas+when+only+one+must+validate.%22&type=Code

**Comment by aeschli** [comment]
Do you have a suggestion on how the message should be improved? IMO the current message is correct.

**Comment by apupier** [comment]
> the current message is correct.

yes, it is correct. Unfortunately, i wasn't able to understand what was wrong and how to fix that. Given the large amount of people asking the same questions on StackOverflow, i think it will be good to find a way to drive more newbies like me.

> Do you have a suggestion on how the message should be improved?

* `Matches multiple schemas when only one must validate. See json schema specification for oneOf http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.2.1.3`
* `The json schema <xxx> is requiring a single schema match as it is using the oneOf constraint. It is matching several schemas. See also json schema specification for oneOf http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.2.1.3`
* `The json schema <xxx> is requiring a single schema match as it is using the oneOf constraint. It is matching several schemas: <schema-1>, <schema-2>, <schema-3>. See also json schema specification for oneOf http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.2.1.3`

**Comment by aeschli** [comment]
I think the problem is that an end user sees the message, but the user doesn't care about schema. The user probably wants a message like `Object can not have property 'x' as well as 'y'.`.

**Comment by vishalsatam** [comment]
Can we allow for a new errorMessage interface that let's you customize the error messages as necessary. The `ajv-errors` library provides something like this https://www.npmjs.com/package/ajv-errors/v/1.0.1 ... but it would be nice to have this natively supported within vscode-json-languageservice
